### DISCUSSION TITLE
Allow some properties of vue-query to accept getter functions

### DISCUSSION THREAD

**Discussion by Mini-ghost** [TanStack/query#5990] [Ideas]
We often need to determine the value of `enabled` based on different conditions, maybe we can allow passing a getters in certain properties?

like this:
```ts
useQuery({
  queryKey: ['TODO'],
  queryFn: () =>  Promise.resolve({ data: []}),

  // Allow `MaybeRefOrGetter<boolean | undefined>`
  enabled() {
    return isMounted.value && isAuth.value
  },
})
```

This was discussed in [Support Vue getters #5461](https://github.com/TanStack/query/discussions/5461), but here the focus is on specific attributes, such as the mentioned `enabled`.

**Comment by DamianOsipiuk** [comment]
Some of the options already support callback functions which are basically getters+. They are running just in time with additional context props.

As for the others, there are a couple of questions
1. How to handle potential conflicts whenever `query-core` adds a callback possibility to a particular property?
2. `getters` are a Vue 3.3 feature. What about Vue 2 compatibility? We would either need a polyfill or return different types for different vue versions.
3. Do you see other properties that would benefit from being getters?

**Reply by Mini-ghost** [reply]
To address the second question first:

### `getters` are a Vue 3.3 feature. What about Vue 2 compatibility? We would either need a polyfill or return different types for different vue versions.

Perhaps we can internally implement "toValue" like this:

```ts
function toValue<T>(r: MaybeRefOrGetter<T>): T {
  return typeof r === 'function'
    ? (r as AnyFn)()
    : unref(r)
}
```

This code is copied from vueuse and works with Vue 2.7 and Vue 3.2 versions.

### Do you see other properties that would benefit from being getters?

I'm not sure. Currently, I always use `enabled` for `computed` properties, and perhaps `queryKey` could be a getter, but it doesn't have an obvious benefit in my case.

This might require further discussion within the community.

### How to handle potential conflicts whenever `query-core` adds a callback possibility to a particular property?

I'm not entirely sure if I understand the question correctly. I'll attempt to discuss it based on my understanding:

Similar to the current `parseQueryArgs`, we will parse all arguments before creating an instance of `QueryObserver`. We can determine whether a specific property is a getter in `parseQueryArgs`, and if it is, execute it. This approach will also have the effect of tracking reaction data.

Take `enabled` as an example:

> [!NOTE]
> This example is updated

```ts
function parseQueryArgs(arg1, arg2 = {}, arg3 = {}) {
  const plainArg1 = unref(arg1)
  const plainArg2 = unref(arg2)
  const plainArg3 = unref(arg3)

  let options = plainArg1

  if (!isQueryKey(plainArg1)) {
    options = plainArg1
  } else if (typeof plainArg2 === 'function') {
    options = { ...plainArg3, queryKey: plainArg1, queryFn: plainArg2 }
  } else {
    options = { ...plainArg2, queryKey: plainArg1 }
  }
  
  const clonedOptions = cloneDeepUnref(options) 

  if (typeof clonedOptions.enabled === 'function') {
    clonedOptions.enabled = clonedOptions.enabled()
  }

  return clonedOptions
}
```

**Reply by DamianOsipiuk** [reply]
I might be missing something, but how would you `watch` if dependencies encapsulated by the `getter` changes? Based on the beta branch.
https://github.com/TanStack/query/blob/beta/packages/vue-query/src/useBaseQuery.ts#L78

As I understand it, we would have to make an explicit watch on every `getter` in options?

**Reply by Mini-ghost** [reply]
`cloneDeepUnref` will only copy arrays and objects and unwrap `ref`, functions (such as `queryFn`) will not be processed.

So we just need to pick the property of type `MaybeRefOrGetter` and call it, like below:

```ts
const defaultedOptions = computed(() => {
  const clonedOptions = cloneDeepUnref(options)

  if (typeof clonedOptions.enabled === 'function') {
    clonedOptions.enabled = clonedOptions.enabled()
  }

  const defaulted = client.defaultQueryOptions(clonedOptions)

  defaulted._optimisticResults = client.isRestoring.value
    ? 'isRestoring'
    : 'optimistic'

  return defaulted
})
```

**Reply by DamianOsipiuk** [reply]
But my question is, how `computed` would know that `enabled` should be evaluated, since `enabled` fn reference would not change?
With other props `computed` can see that there is a ref in options and reevaluate when it changes, but `getter` is just a function.
Considering no other option changes, but only a ref that is used only inside `enabled getter`.

If you could provide a test with this that will work on beta branch against Vue 2 and react to getter dependency change we could think to integrate that.

**Reply by Mini-ghost** [reply]
The following example demonstrates that Vue 2 can support the getters API.
https://stackblitz.com/edit/vitejs-vite-9rq3ru?file=src%2FApp.vue

This is a PR for the `master` branch

- https://github.com/TanStack/query/pull/6004

I seem to have misunderstood what you meant, maybe I should prepare a PR for the `beta` branch?

**Reply by Mini-ghost** [reply]
This PR for the `beta` branch:

- https://github.com/TanStack/query/pull/6018

**Comment by suneettipirneni** [comment]
Just in case anyone else stumbles upon this discussion, getters are now supported [for the `enabled` option](https://github.com/TanStack/query/pull/6018) as a well for [`queryKey`s](https://github.com/TanStack/query/issues/7608)
